# Analysis for Hacker News Posts
[Hacker](https://news.ycombinator.com/) News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.  
You can find the data set [here](https://www.kaggle.com/hacker-news/hacker-news-posts), but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions. Below are descriptions of the columns:

**id**: The unique identifier from Hacker News for the post  

**title**: The title of the post  

**url**: The URL that the posts links to, if it the post has a URL  

**num_points**: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes  

**num_comments**: The number of comments that were made on the post 

**author**: The username of the person who submitted the post  

**created_at**: The date and time at which the post was submitted  

We'll compare these two types of posts to determine the following:
 - Do Ask HN (posts to ask the Hacker News community a specific question) or Show HN (posts to show the Hacker News community a project, product, or just generally something interesting) receive more comments on average?
 - Do posts created at a certain time receive more comments on average?
 
First, let's open our file

In [1]:
opened_file = open("hacker_news.csv")
from csv import reader
read_file= reader(opened_file)
hn = list(read_file)
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


Now, we can remove the header row, to keep only our data that we will work with.

In [2]:
header = hn[0]
hn = hn[1:]
print(header)
print()
print (hn[0])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']


## Extracting Ask HN and Show HN Posts
Now to start our analysis , we need to work with titles to separate our data into two main lists, Ask HN and Show HN, not to mention a third list containing all other posts. We will loop through our data to get these titles.

In [3]:
ask_posts = []
show_posts = []
other_posts = []
for row in hn:
    title = row[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))
        

1744
1162
17194


## Calculating the Average Number of Comments for Ask HN and Show HN Posts

In [4]:
total_ask_comments = 0
for row in ask_posts:
    total_ask_comments += int(row[4])
print(total_ask_comments)

24483


In [5]:
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [6]:
total_show_comments = 0
for row in show_posts:
    total_show_comments += int(row[4])
avg_show_posts = total_show_comments / len(show_posts)
print(avg_show_posts)

10.31669535283993


As we can see in the above code, we calculated the average number of comments a posts get in the two lists we have. Our finding show that Ask posts get on average approximately 14 comments per post. On the other hand Show posts get on average approximately 10 comments per post.

**Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts**

## Finding the Amount of Ask Posts and Comments by Hour Created

The amount of ask posts created in each hour of the day, along with the number of comments received.

In [55]:
import datetime as dt
# result list to store the time of creation and the number of comments of each post 
result_list = []
for row in ask_posts:
    creat_at = row[6]
    n_comments = int(row[4])
    result_list.append([creat_at, n_comments])

counts_by_hour = {}
comments_by_hour= {}

for row in result_list:
    time = row[0]
    comment = row[1]
    hour1 = dt.datetime.strptime(time, "%m/%d/%Y %H:%M")
    hour = dt.datetime.strftime(hour1, "%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comment       
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comment
        
counts_by_hour

{'00': 55,
 '01': 60,
 '02': 58,
 '03': 54,
 '04': 47,
 '05': 46,
 '06': 44,
 '07': 34,
 '08': 48,
 '09': 45,
 '10': 59,
 '11': 58,
 '12': 73,
 '13': 85,
 '14': 107,
 '15': 116,
 '16': 108,
 '17': 100,
 '18': 109,
 '19': 110,
 '20': 80,
 '21': 109,
 '22': 71,
 '23': 68}

In [57]:
comments_by_hour

{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

The average number of comments Ask HN posts receive by hour

In [61]:
avg_by_hour = []
for h in comments_by_hour:
    avg_by_hour.append([h, comments_by_hour[h]/counts_by_hour[h]])
avg_by_hour

[['23', 7.985294117647059],
 ['07', 7.852941176470588],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['15', 38.5948275862069],
 ['13', 14.741176470588234],
 ['12', 9.41095890410959],
 ['02', 23.810344827586206],
 ['14', 13.233644859813085],
 ['06', 9.022727272727273],
 ['04', 7.170212765957447],
 ['09', 5.5777777777777775],
 ['08', 10.25],
 ['05', 10.08695652173913],
 ['10', 13.440677966101696],
 ['17', 11.46],
 ['22', 6.746478873239437],
 ['21', 16.009174311926607],
 ['00', 8.127272727272727],
 ['20', 21.525],
 ['11', 11.051724137931034],
 ['16', 16.796296296296298],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297]]

## Sorting and Printing Values from a List of Lists

In [66]:
swap_avg_by_hour= []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
swap_avg_by_hour

[[7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [10.8, '19'],
 [11.383333333333333, '01'],
 [38.5948275862069, '15'],
 [14.741176470588234, '13'],
 [9.41095890410959, '12'],
 [23.810344827586206, '02'],
 [13.233644859813085, '14'],
 [9.022727272727273, '06'],
 [7.170212765957447, '04'],
 [5.5777777777777775, '09'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [13.440677966101696, '10'],
 [11.46, '17'],
 [6.746478873239437, '22'],
 [16.009174311926607, '21'],
 [8.127272727272727, '00'],
 [21.525, '20'],
 [11.051724137931034, '11'],
 [16.796296296296298, '16'],
 [13.20183486238532, '18'],
 [7.796296296296297, '03']]

In [69]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [74]:
print("Top 5 Hours for Ask Posts Comments")
top_5 = sorted_swap[:5]
top_5

Top 5 Hours for Ask Posts Comments


[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21']]

In [87]:
for avg, hr in top_5:
    final = dt.datetime.strptime(hr,'%H').strftime('%H:%M')
    final1 = "{}: {:.2f}  average comments per post ".format(final,avg)
    print(final1)

15:00: 38.59  average comments per post 
02:00: 23.81  average comments per post 
20:00: 21.52  average comments per post 
16:00: 16.80  average comments per post 
21:00: 16.01  average comments per post 


As we can see in the results displayed above, the hour 15:00 (3 PM) receives the most average comments during the day at 38.59 comments. 

## Conclusion

After working though all the data in the Hackers News website, our results show that generally Ask HN posts reveives more comments in average than Show HN posts do. Then after analyzing the Ask posts, we conclude that Ask posts that received comments are more likely to receive the most posts around the hour of 15:00 which transelate into 3:00 PM est.  
As a result, we recommand to post around 3 PM Eastern Time Zone.

# Extra Work

In [93]:
# average points per ask posts
total_ask_points= 0
for row in ask_posts:
    total_ask_points += int(row[3])
print(total_ask_points)
avg_ask_points = total_ask_points / len(ask_posts)
avg_ask_points

26268


15.061926605504587

In [95]:
# average points per show posts
total_show_points= 0
for row in show_posts:
    total_show_points += int(row[3])
print(total_show_points)
avg_show_points = total_show_points / len(show_posts)
avg_show_points

32019


27.555077452667813